### PR TITLE
Implemented accumulate for recursive folding

### PR DESCRIPTION
**Pull request by michaelquinn32**
Following up on #46 and #11, I implemented a version of `accumulate`, which adds recursive folding to the package. 

The differences between `accumulate` and `reduce` are minor. I simply added another argument (`accumulate = TRUE`) to the call of `Reduce`. Nonetheless, this change provides a lot of utility. 

I have two examples in the documentation. The first shows flooring and capping a dataset, which is an extension of an example in [Haskell's scanl documentation](http://zvon.org/other/haskell/Outputprelude/scanl_f.html). The second example shows a simple simulation of a random walk with drift.

### CODE REVIEW

File: R/reduce.R [hunk 1]
```diff
@@ -46,3 +46,54 @@ reduce_right <- function(.x, .f, ..., .init) {
...
+#'   map2(caps, ~ accumulate(.x, min, .init = .y[2]))
+#'
+#' # Simulating stochastic processes with drift
+#' \dontrun{
+#' library(dplyr)
+#' library(ggplot2)
+#'
+#' rerun(5, rnorm(100)) %>% setNames(paste0("sim", 1:5)) %>%
+#'   map(~ accumulate(., function(.x, .y) .05 + .x + .y)) %>%
+#'   map(~ data.frame(value = .x, step = 1:100)) %>%
```
Comment (hadley) [inline R/reduce.R]: This could now be `map_df(~ data_frame(value = .x, step = 1:100), .id = "simulation")`

File: R/reduce.R [hunk 2]
```diff
@@ -46,3 +46,54 @@ reduce_right <- function(.x, .f, ..., .init) {
...
+#'   bind_rows(. , .id = "simulation") %>%
+#'   ggplot(., aes(x = step, y = value)) + geom_line(aes(color = simulation))
```
Comment (hadley) [inline R/reduce.R]: You don't need the explicit `.` here

File: R/reduce.R [hunk 3]
```diff
@@ -46,3 +46,54 @@ reduce_right <- function(.x, .f, ..., .init) {
...
+#' caps <- mtcars %>% map(~ quantile(.x, probs = c(.1, .9), names = FALSE))
+#' floor_cap <- mtcars %>% map2(caps, ~ accumulate(.x, max, .init = .y[1])) %>%
+#'   map2(caps, ~ accumulate(.x, min, .init = .y[2]))
+#'
+#' # Simulating stochastic processes with drift
+#' \dontrun{
+#' library(dplyr)
+#' library(ggplot2)
+#'
+#' rerun(5, rnorm(100)) %>% setNames(paste0("sim", 1:5)) %>%
```
Comment (hadley) [inline R/reduce.R]: Can you put this on a new line and use `set_names()`?

File: R/reduce.R [hunk 4]
```diff
@@ -46,3 +46,54 @@ reduce_right <- function(.x, .f, ..., .init) {
...
+#'   map(~ accumulate(., function(.x, .y) .05 + .x + .y)) %>%
+#'   map(~ data.frame(value = .x, step = 1:100)) %>%
+#'   bind_rows(. , .id = "simulation") %>%
+#'   ggplot(., aes(x = step, y = value)) + geom_line(aes(color = simulation))
+#' }
+accumulate <- function(.x, .f, ..., .init) {
+  force(.f)
```
Comment (hadley) [inline R/reduce.R]: I think you should use `as_function()` here to be consistent with the rest of purrr.

**Comment by hadley** [comment]
Looks good - thanks!  Can you please respond to my super nitpicky comments and add a bullet to NEWS?

**Comment by michaelquinn32** [comment]
Super nitpicky is super helpful! I didn't notice the addition of `map_df`. That's a great little tool. 

I also did a little additional formatting to the examples. Hopefully they'll be a little easier to read.

Please let me know if I missed anything.

File: R/reduce.R (hunk 4 above)
Comment (michaelquinn32) [inline R/reduce.R]: I kept the syntax for `accumulate()` as close as possible to that of `reduce()`. Does it make sense to use `as_function()` in one but not the other? Or should I just change both?

**Comment by hadley** [comment]
Looks great - thanks!

**Pull request merged**